eventstream/eventstream.go

```diff
@@ -21,61 +26,34 @@ func Publish(event interface{}) {
 	es.Publish(event)
 }
 
-func PublishUnsafe(evt interface{}) {
-	es.PublishUnsafe(evt)
-}
-
 type EventStream struct {
 	sync.RWMutex
-	subscriptions []*Subscription
+	subscriptions cmap.ConcurrentMap
 }
 
 func (es *EventStream) Subscribe(fn func(evt interface{})) *Subscription {
-	es.Lock()
+	uuid := uuid.New().String()
 	sub := &Subscription{
 		es: es,
-		i:  len(es.subscriptions),
+		id: uuid,
 		fn: fn,
 	}
-	es.subscriptions = append(es.subscriptions, sub)
-	es.Unlock()
+	es.subscriptions.SetIfAbsent(uuid, sub)
 	return sub
 }
 
 func (es *EventStream) Unsubscribe(sub *Subscription) {
-	if sub.i == -1 {
-		return
-	}
-
-	es.Lock()
-	i := sub.i
-	l := len(es.subscriptions) - 1
-
-	es.subscriptions[i] = es.subscriptions[l]
-	es.subscriptions[i].i = i
-	es.subscriptions[l] = nil
-	es.subscriptions = es.subscriptions[:l]
-	sub.i = -1
-
-	// TODO(SGC): implement resizing
-	if len(es.subscriptions) == 0 {
-		es.subscriptions = nil
-	}
-
-	es.Unlock()
+	es.subscriptions.Remove(sub.id)
 }
 
 func (es *EventStream) Publish(evt interface{}) {
-	es.RLock()
-	defer es.RUnlock()
-
-	es.PublishUnsafe(evt)
-}
-
-func (es *EventStream) PublishUnsafe(evt interface{}) {
-	for _, s := range es.subscriptions {
-		if s.p == nil || s.p(evt) {
-			s.fn(evt)
+	for _, key := range es.subscriptions.Keys() {
+		i, ok := es.subscriptions.Get(key)
+		if ok {
+			s, ok := i.(*Subscription)
+			if ok && (s.p == nil || s.p(evt)) {
```

Comment:
I'm not entirely sure about this:
Imagine `s.p == nil` evaluates to false, `WithPredicate(nil)` is called and then `s.p(evt)` is evaluated. This would result in an NPE.

Hence I don't think it is save to access `s.p` here without any locking mechanism.